VocaDbWeb.Core/Controllers/StatsController.cs

```diff
@@ -0,0 +1,805 @@
+using VocaDb.Model.Domain.PVs;
+using VocaDb.Model.Domain.Security;
+using VocaDb.Model.Domain.Songs;
+using VocaDb.Model.Domain.Tags;
+using VocaDb.Model.Domain.Users;
+using VocaDb.Model.Service.QueryableExtensions;
+using VocaDb.Model.Utils.Config;
+using VocaDb.Web.Code.Highcharts;
+using VocaDb.Web.Helpers;
+
+namespace VocaDb.Web.Controllers
+{
+	public class StatsController : ControllerBase
+	{
+		private const int ClientCacheDurationSec = 86400;
+		private readonly VdbConfigManager _config;
+		private readonly ActivityEntryQueries _activityEntryQueries;
+		private readonly StatsQueries _queries;
+		private readonly SongAggregateQueries _songAggregateQueries;
+
+		// TODO: implement
+		/*private T GetCachedReport<T>() where T : class
+		{
+			var name = ControllerContext.RouteData.Values["action"] + "_" + ControllerContext.RouteData.Values["cutoff"];
+			var item = _context.Cache["report_" + name];
+
+			if (item == null)
+				return null;
+
+			return (T)item;
+		}*/
+		private T GetCachedReport<T>() where T : class => throw new NotImplementedException();
+
+		// TODO: implement
+		/*private void SaveCachedReport<T>(T data) where T : class
+		{
+			var name = ControllerContext.RouteData.Values["action"];
+			_context.Cache.Add("report_" + name, data, null, Cache.NoAbsoluteExpiration, TimeSpan.FromDays(1), CacheItemPriority.Default, null);
+		}*/
+		private void SaveCachedReport<T>(T data) where T : class => throw new NotImplementedException();
+
+		private ActionResult AreaChart(string title, params Series[] dataSeries)
+		{
+			var json = new Highchart
+			{
+				Chart = new Chart
+				{
+					Height = 600,
+					Type = ChartType.Area
+				},
+				Title = title,
+				XAxis = new Axis(AxisType.Datetime, new Title()),
+				YAxis = new Axis
+				{
+					Title = "Percentage",
+					Min = 0
+				},
+				Tooltip = new
+				{
+					Shared = true,
+					Crosshairs = true
+				},
+				PlotOptions = new PlotOptions
+				{
+					Bar = new
+					{
+						DataLabels = new
+						{
+							Enabled = true
+						}
+					},
+					Area = new PlotOptionsArea
+					{
+						Stacking = PlotOptionsAreaStacking.Percent,
+						LineColor = "#ffffff",
+						LineWidth = 1,
+						Marker = new
+						{
+							LineWidth = 1,
+							LineColor = "#ffffff"
+						}
+					}
+				},
+				Legend = new
+				{
+					Layout = "vertical",
+					Align = "left",
+					X = 120,
+					VerticalAlign = "top",
+					Y = 100,
+					Floating = true,
+					BackgroundColor = "#FFFFFF"
+				},
+				Series = (
+					dataSeries
+				)
+			};
+
+			return LowercaseJson(json);
+		}
+
+		private ActionResult DateLineChartWithAverage(string title, string pointsTitle, string yAxisTitle, ICollection<Tuple<DateTime, int>> points,
+			bool average = true)
+		{
+			Response.GetTypedHeaders().CacheControl = new CacheControlHeaderValue
+			{
+				Public = true,
+				MaxAge = TimeSpan.FromDays(1),
+			};
+			// TODO: implement Response.Cache.SetSlidingExpiration(true);

```

Comment:
How can I perform sliding expiration?

Reply:
It's not that important. Max age is enough here.